bootloader/src/old_machine_bootloader_linux.c

```diff
@@ -0,0 +1,290 @@
+    size_t to_be_replaced_size = strlen(to_be_replaced);
+    size_t replacement_size = strlen(replacement);
+    size_t result_size = str_size - to_be_replaced_size + replacement_size;
+    char *result_string = (char*)malloc(sizeof(char) * (result_size));
+    bool substringReplaced = false;
+
+    for (int i = 0; i < (int)result_size; i++ ){
+        if(substringReplaced){
+            break;
+        }
+        result_string[i] = str[i];
+        if(str[i] == to_be_replaced[0] && replacement_size != 0){
+            bool should_replace = true;
+            // Check if started iterating over string that will be replaced
+            for (int j = i; j < (i + to_be_replaced_size); j++){
+                if(str[j] != to_be_replaced[j - i]) {
+                    should_replace = false;
+                }
+            }
+            // If string that needs to be replaced is found - replace it
+            if (should_replace) {
+                for (int j = i; j < (i + replacement_size); j++){
+                    result_string[j] = replacement[j - i];
+                }
+                substringReplaced = true;
+            }
+        }
+    }
+    result_string[result_size] = '\0';
+    if(substringReplaced){
+        return result_string;
+    } else {
+        return str;
+    }
+}
+
+char* concatenate(size_t size, char *array[size], const char *joint){
+    size_t jlen, lens[size];
+    size_t i, total_size = (size-1) * (jlen=strlen(joint)) + 1;
+    char *result, *p;
+    for(i=0;i<size;++i){
+        total_size += (lens[i]=strlen(array[i]));
+    }
+    p = result = malloc(total_size);
+    for(i=0;i<size;++i){
+        printf("%s\n", array[i]);
+        memcpy(p, array[i], lens[i]);
+        p += lens[i];
+        if(i<size-1){
+            memcpy(p, joint, jlen);
+            p += jlen;
+        }
+    }
+    *p = '\0';
+    return result;
+}
+
+struct response sendRequest(char* server, char* tunnel, bool tunnelUsed, char* data){
+    CURL *curl;
+    CURLcode res;
+    struct response s;
+    curl = curl_easy_init();
+    init_response(&s);
+    if(curl) {
+        curl_easy_setopt(curl, CURLOPT_URL, server);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writefunc);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &s);
+        if(tunnelUsed){
+            curl_easy_setopt(curl, CURLOPT_PROXY, tunnel);
+        }
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
+        /* Perform the request, res will get the return code */
+        res = curl_easy_perform(curl);
+        /* Check for errors */
+        if(res != CURLE_OK){
+            fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+            s.ptr = "FAILED";
+            s.len = strlen(s.ptr);
+            return s;
+        }
+
+        /* always cleanup */
+        curl_easy_cleanup(curl);
+    }
+    curl_global_cleanup();
+    return s;
+}
+
+char** getIpAddresses(int maxSize, int *addrCount){
```

Comment:
Why do you need maxSize? Figure out how many IPs exist on the machine then allocate the correct amount